engine/src/main/java/org/apache/apex/engine/YarnAppLauncherImpl.java

```diff
@@ -164,8 +160,23 @@ public void shutdown(org.apache.apex.api.Launcher.ShutdownMode shutdownMode)
         throws org.apache.apex.api.Launcher.LauncherException
     {
       shutdownApp(this, shutdownMode);
+    }
 
+    @Override
+    public void close() throws Exception
+    {
+      yarnClient.close();
+      yarnClient = null;
     }
 
+    // No guarantee that shutdown or close will be called to do a final cleanup
+    // So closing yarnClient when the memory is recycled
+    @Override
+    protected void finalize() throws Throwable
+    {
```

Comment:
I am not sure that this is necessary as YarnClient or YarnClientImpl does not overwrite finalize(). Additionally, using finalize leads to checkstyle violation and needs to be fixed. Not calling shutdown or close is misuse of the class. Should you decide to keep it, finalize should log warning requesting a fix (proper call to close()).